src/modules/mc_raptor/mc_raptor.cpp

```diff
@@ -0,0 +1,1078 @@
+		updateArmingCheckReply();
+		return;
+	}
+
+	if ((current_time - timestamp_last_angular_velocity) > OBSERVATION_TIMEOUT_ANGULAR_VELOCITY) {
+		status.exit_reason = raptor_status_s::EXIT_REASON_ANGULAR_VELOCITY_STALE;
+
+		if constexpr(PUBLISH_NON_COMPLETE_STATUS) {
+			_raptor_status_pub.publish(status);
+		}
+
+		if (!timeout_message_sent) {
+			PX4_ERR("angular velocity timeout");
+			timeout_message_sent = true;
+		}
+
+		can_arm = false;
+		updateArmingCheckReply();
+		return;
+	}
+
+	if ((current_time - timestamp_last_local_position) > OBSERVATION_TIMEOUT_LOCAL_POSITION) {
+		status.exit_reason = raptor_status_s::EXIT_REASON_LOCAL_POSITION_STALE;
+
+		if constexpr(PUBLISH_NON_COMPLETE_STATUS) {
+			_raptor_status_pub.publish(status);
+		}
+
+		if (!timeout_message_sent) {
+			PX4_ERR("local position timeout");
+			timeout_message_sent = true;
+		}
+
+		can_arm = false;
+		updateArmingCheckReply();
+		return;
+
+	} else {
+		position[0] = _vehicle_local_position.x;
+		position[1] = _vehicle_local_position.y;
+		position[2] = _vehicle_local_position.z;
+		linear_velocity[0] = _vehicle_local_position.vx;
+		linear_velocity[1] = _vehicle_local_position.vy;
+		linear_velocity[2] = _vehicle_local_position.vz;
+	}
+
+	// position and linear_velocity are guaranteed to be set after this point
+	auto previous_internal_reference = internal_reference;
+	internal_reference = static_cast<InternalReference>(_param_mc_raptor_intref.get());
+	bool internal_reference_changed = previous_internal_reference != internal_reference;
+
+	if (internal_reference_changed) {
+		PX4_INFO("internal reference changed from %d to %d", (int)previous_internal_reference, (int)internal_reference);
+	}
+
+	if (use_internal_reference && internal_reference != InternalReference::NONE) {
+		if (next_active && (!previous_active || internal_reference_changed || internal_reference_params_changed)) {
+			internal_reference_activation_position[0] = position[0];
+			internal_reference_activation_position[1] = - position[1];
+			internal_reference_activation_position[2] = - position[2];
+			internal_reference_activation_orientation[0] = _vehicle_attitude.q[0];
+			internal_reference_activation_orientation[1] = _vehicle_attitude.q[1];
+			internal_reference_activation_orientation[2] = -_vehicle_attitude.q[2];
+			internal_reference_activation_orientation[3] = -_vehicle_attitude.q[3];
+			internal_reference_activation_time = current_time;
+			PX4_INFO("internal reference activated at: %f %f %f", (double)internal_reference_activation_position[0],
+				 (double)internal_reference_activation_position[1], (double)internal_reference_activation_position[2]);
+			internal_reference_params_changed = false;
+		}
+
+		Setpoint setpoint{};
+
+		if (internal_reference == InternalReference::LISSAJOUS) {
+			setpoint = lissajous(static_cast<T>(current_time - internal_reference_activation_time) / 1000000, lissajous_params);
+
+		} else {
+			PX4_ERR("internal reference type not supported");
```

Comment:
What would be the missing pieces on using other trajectories OR simply a reference instead?